Tests: Creating unittests for the hash_string module

kushalsingh007 opened the pull request:
- Created unittests for hash_string module present in sys/hash_string
- Fixes a part of #1131
- Created a new PR because of some problems in my previous branch for #2698

kushalsingh007 commented:
@Kijewski : I hope it solves the points you made in PR #2698

Kijewski commented on sys/hash_string/hash_string.c:
```diff
@@ -26,7 +26,7 @@ unsigned long hash_string(unsigned char *str)
     unsigned long hash = 5381;
     int c;
 
-    while ((c = *str++)) {
```
The extra parens are intentionally here, because using `=` in a condition is almost always a typo.
I'd prefer `for (char c; (c = *str) != 0; ++str)`, though.

Kijewski commented on tests/unittests/tests-hash_string/tests-hash_string.c:
```diff
@@ -0,0 +1,35 @@
+static void test_hash_value(void)
+{
+    TEST_ASSERT_EQUAL_INT(177621, hash_string("0"));
+    TEST_ASSERT_EQUAL_INT(6555485335635392041, hash_string("0123456789abcde-0123456789abcde-0123456789abcde-0123456789abcde-"));
```
Can you add a linebreak somewhere? You can concat string constants like that in C:
```c
char *str = "abcd" "efgh"; /* no operator in between */
```

kushalsingh007 commented:
@Kijewski : Made the changes you suggested

authmillenon commented on tests/unittests/tests-hash_string/tests-hash_string.c:
```diff
@@ -0,0 +1,40 @@
+{
+    TEST_ASSERT_EQUAL_INT(177621, hash_string("0"));
+    TEST_ASSERT_EQUAL_INT(6555485335635392041, hash_string("0123456789abcde-01"
+                                                           "23456789abcde-0123"
```
Tab in indent.

authmillenon commented on tests/unittests/tests-hash_string/tests-hash_string.c:
```diff
@@ -0,0 +1,40 @@
+    TEST_ASSERT_EQUAL_INT(177621, hash_string("0"));
+    TEST_ASSERT_EQUAL_INT(6555485335635392041, hash_string("0123456789abcde-01"
+                                                           "23456789abcde-0123"
+                                                           "456789abcde-012345"
```
Tab in indent.

authmillenon commented on tests/unittests/tests-hash_string/tests-hash_string.c:
```diff
@@ -0,0 +1,40 @@
+    TEST_ASSERT_EQUAL_INT(6555485335635392041, hash_string("0123456789abcde-01"
+                                                           "23456789abcde-0123"
+                                                           "456789abcde-012345"
+                                                           "6789abcde-"));
```
Tab in indent.

authmillenon commented on tests/unittests/tests-hash_string/tests-hash_string.c:
```diff
@@ -0,0 +1,40 @@
+                                                           "456789abcde-012345"
+                                                           "6789abcde-"));
+    TEST_ASSERT_EQUAL_INT(210674770391, hash_string("Frank jagt im komplett ve"
+                                                    "rwahrlosten Taxi q"
```
Tab in indent.

authmillenon replied:
+ weird syllable division ;-)

authmillenon commented on tests/unittests/tests-hash_string/tests-hash_string.c:
```diff
@@ -0,0 +1,40 @@
+                                                           "6789abcde-"));
+    TEST_ASSERT_EQUAL_INT(210674770391, hash_string("Frank jagt im komplett ve"
+                                                    "rwahrlosten Taxi q"
+                                                    "uer durch Bayern"));
```
Tab in indent.

authmillenon commented:
Please configure your Editor or IDE to use 4 space characters for indentation.

authmillenon commented on sys/hash_string/hash_string.c:
```diff
@@ -26,7 +26,7 @@ unsigned long hash_string(unsigned char *str)
     int c;
 
-    while ((c = *str++)) {
+    for (char c; (c = *str) != 0; ++str) {
```
I'd even would use `'\0'` instead of `0` ;-).

authmillenon replied:
But in general: this change is a cosmetic change that is not really in the scope of this PR, so I would prefer to not add any changes to this file at all.

authmillenon commented on sys/hash_string/hash_string.c:
```diff
@@ -26,7 +26,7 @@ unsigned long hash_string(unsigned char *str)
     unsigned long hash = 5381;
     int c;
```
Not needed anymore with the change below.

kushalsingh007 commented:
@authmillenon : I think I have fixed the changes you suggested , also changed my vim settings.  Also checked with ```git diff --check```. I guess no tab changes left :confused: 
Can I squash it if there are no more changes ( or for changes till now ) ?

authmillenon commented on sys/hash_string/hash_string.c:
```diff
@@ -24,9 +24,8 @@
-    int c;
 
-    while ((c = *str++)) {
+    for (char c; (c = *str) != 0; ++str) {
```
As stated before: this change is neither related to the unittests nor is it presenting any benefits apart from cosmetic ones based on personal taste. I would prefer to have the changes to this file removed from this PR.

authmillenon commented:
Apart from my last comment I'm fine with this PR.

kushalsingh007 commented:
@authmillenon : Done with your suggestion and squashed the previous commits :)

There were to points regarding travis that I wanted to ask
1)Also,Is there a way for a collaborator to stop his build, because for instance this time I had to to ```git push``` two times ( becasue I left some changes the first time) , and because of that even though I know that only my most recent one is correct ,but travis still builds the old one's . So if I can stop my own builds not only would that be faster for everyone , but also it would stop travis from overloading so much :thought_balloon: 
2)If a commit does have ```squash me``` , in it's name wouldn't it be more beneficial to stop the build completely than having only static tests fail and still check run the other testcases ?

authmillenon commented:
Please don't squash until we explicitly say so. Unsquashed commits help us to track your changes and speed up the code review. To your points: yes it would be great if the contributers could control Travis too, but sadly they can't. I try to kill obsoleted builds whenever I see them, but even I have to sleep sometime. I think the best thing you can do right now is not to push everytime you commit something, but only when you are done ;). 
On your second point: I'm personally against it. At least with my PRs I can see that it builds for all boards even before the review process is done, not just on the finish line.

kushalsingh007 commented:
Sorry for being in such a hurry :(  I was under the assumption that if all the changes suggested till that point have been made then one should squash , also I think I took the part when  you said you were fine with the PR as - all the changes have been done. I guess I should read them more closely and be a little more patient. I will keep it in mind the next time.

authmillenon commented:
Travis is happy, no squashing is needed anymore, so ACK & go ;-)
